### PR TITLE
wip: Token conditional swaps

### PR DESCRIPTION
**Pull request by ckamm**

### CODE REVIEW

File: ts/client/src/client.ts [hunk 1]
```diff
@@ -810,86 +811,47 @@ export class MangoClient {
...
-      }
-    }
+    // Re-encode decoded mango account with v1 layout, this will help identifying
+    // if account is of type v1 or v2
+    // Do whole encoding manually, since anchor uses a buffer of a constant length which is too small
+    let mangoAccountV1Buffer = Buffer.alloc(ai.data.length);
+    const layout =
+      this.program.coder.accounts['accountLayouts'].get('mangoAccount');
+    const len = layout.encode(decodedMangoAccount, mangoAccountV1Buffer);
+    mangoAccountV1Buffer = mangoAccountV1Buffer.subarray(0, len);
```
Comment (ckamm) [inline ts/client/src/client.ts:838]: Since the len is all we care about, we could skip getting the buffer and just compare to len + 8

File: ts/client/src/client.ts [hunk 2]
```diff
@@ -810,86 +811,47 @@ export class MangoClient {
...
+    const discriminator =
+      BorshAccountsCoder.accountDiscriminator('mangoAccount');
+    mangoAccountV1Buffer = Buffer.concat([discriminator, mangoAccountV1Buffer]);
+
+    const tokenConditionalSwaps =
+      mangoAccountV1Buffer.length < ai.data.length
+        ? (borsh
```
Comment (ckamm) [inline ts/client/src/client.ts:845]: would prefer this to be a full `if`

File: ts/client/src/client.ts [hunk 3]
```diff
@@ -937,44 +902,45 @@ export class MangoClient {
...
-                bytes: anchroHeaderFilter.bytes,
-                offset: anchroHeaderFilter.offset,
+    const accounts = await Promise.all(
+      (
+        await this.program.provider.connection.getProgramAccounts(
+          this.programId,
+          {
+            filters: [
+              {
+                memcmp: {
```
Comment (ckamm) [inline ts/client/src/client.ts:924]: memcmp: discriminatorMemcmp didn't do it?

Comment (microwavedcola1) [reply ts/client/src/client.ts:924]: sorry, could you rephrase please?

File: ts/client/src/client.ts (hunk 2 above)
Comment (microwavedcola1) [reply ts/client/src/client.ts:845]: js makes it hard with full ifs, then tokenConditionalSwaps needs to be defined separately with a let, and then reassigned, keeping as is

File: programs/mango-v4/src/instructions/account_create.rs [hunk 4]
```diff
@@ -29,7 +29,7 @@ pub fn account_create(
     account.fixed.delegate = Pubkey::default();
     account.fixed.set_being_liquidated(false);
 
-    account.expand_dynamic_content(token_count, serum3_count, perp_count, perp_oo_count)?;
+    account.expand_dynamic_content(token_count, serum3_count, perp_count, perp_oo_count, 0)?;
```
Comment (microwavedcola1) [inline programs/mango-v4/src/instructions/account_create.rs:32]: maybe we should start with a default of 4?

File: programs/mango-v4/src/instructions/group_edit.rs [hunk 5]
```diff
@@ -106,5 +108,24 @@ pub fn group_edit(
...
+        require_gte!(fees_bps, 0); // values <0 are not currently supported
+        group.token_conditional_swap_taker_fee_bps = fees_bps;
+    }
+    if let Some(fees_bps) = token_conditional_swap_maker_fee_bps_opt {
+        msg!(
+            "Token conditional swap maker fee bps old {:?}, new {:?}",
+            group.token_conditional_swap_maker_fee_bps,
+            fees_bps
+        );
+        require_gte!(fees_bps, 0); // values <0 are not currently supported
```
Comment (microwavedcola1) [inline programs/mango-v4/src/instructions/group_edit.rs:126]: hmm why dont we want to support -ve maker fees?

File: programs/mango-v4/src/state/mango_account.rs [hunk 6]
```diff
@@ -629,6 +714,41 @@ impl<
...
+    pub fn token_conditional_swap_by_index(&self, index: usize) -> Result<&TokenConditionalSwap> {
+        require_gt!(self.header().token_conditional_swap_count(), index);
+        Ok(self.token_conditional_swap_by_index_unchecked(index))
+    }
+
+    pub fn token_conditional_swap_by_id(&self, id: u64) -> Result<(usize, &TokenConditionalSwap)> {
+        let index = self
+            .all_token_conditional_swap()
+            .position(|tcs| tcs.is_active() && tcs.id == id)
+            .ok_or_else(|| error_msg!("token stop loss with id {} not found", id))?;
```
Comment (microwavedcola1) [inline programs/mango-v4/src/state/mango_account.rs:733]: this is the very few place where we directly say "stop loss", nit: maybe reword to token conditional swap with ...?

File: programs/mango-v4/src/state/token_conditional_swap.rs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    /// maximum amount of native tokens to buy or sell
+    pub max_buy: u64,
+    pub max_sell: u64,
+
+    /// how many native tokens were already bought/sold
+    pub bought: u64,
+    pub sold: u64,
+
+    /// timestamp until which the conditional swap is valid
+    pub expiry_timestamp: u64,
```
Comment (microwavedcola1) [inline programs/mango-v4/src/state/token_conditional_swap.rs:43]: maybe is_active method everywhere should also use this field to mark as active, not active or free tcs slot

File: programs/mango-v4/src/instructions/token_conditional_swap_trigger.rs [hunk 8]
```diff
@@ -0,0 +1,705 @@
...
+    require!(tcs.is_active(), MangoError::SomeError);
+    require_eq!(tcs.id, token_conditional_swap_id);
+
+    // Possibly wipe the tcs and exit, if it's already expired
+    let now_ts: u64 = Clock::get()?.unix_timestamp.try_into().unwrap();
+    if tcs.is_expired(now_ts) {
+        drop(tcs);
+        let tcs = liqee.token_conditional_swap_mut_by_index(token_conditional_swap_index)?;
+        *tcs = TokenConditionalSwap::default();
+
```
Comment (microwavedcola1) [inline programs/mango-v4/src/instructions/token_conditional_swap_trigger.rs:64]: a msg saying already expired would be useful here

File: ts/client/src/client.ts (hunk 3 above)
Comment (ckamm) [reply ts/client/src/client.ts:921]: I mean, do you need to do
```
memcmp: {
    bytes: a.bytes,
    offset: a.offset,
}
```
? wouldn't `memcmp: a` work?

File: programs/mango-v4/src/instructions/account_create.rs (hunk 4 above)
Comment (ckamm) [reply programs/mango-v4/src/instructions/account_create.rs:32]: yeah, some default would be good. AccountCreateV2 I think, so it can be ui-controlled, like the others

File: programs/mango-v4/src/instructions/group_edit.rs (hunk 5 above)
Comment (ckamm) [reply programs/mango-v4/src/instructions/group_edit.rs:126]: It's just something I haven't thought about yet. Even though these makers are technically providing liquidity, it's not really a cohesive market. Also, we already need to incentivize the takers to execute these, so it feels odd to shift funds from takers to makers somehow.

File: programs/mango-v4/src/state/mango_account.rs (hunk 6 above)
Comment (ckamm) [reply programs/mango-v4/src/state/mango_account.rs:733]: yeah, this is a leftover, will fix

File: programs/mango-v4/src/state/token_conditional_swap.rs (hunk 7 above)
Comment (ckamm) [reply programs/mango-v4/src/state/token_conditional_swap.rs:43]: maybe I'll rename `is_active()` to `has_data` or something

File: programs/mango-v4/src/state/token_conditional_swap.rs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    /// Premium price adjusted for the maker fee
+    pub fn maker_price(&self, premium_price: f32) -> f32 {
+        premium_price * (1.0 + self.maker_fee_bps as f32 * 0.0001)
+    }
+
+    /// Premium price adjusted for the taker fee
+    pub fn taker_price(&self, premium_price: f32) -> f32 {
+        premium_price * (1.0 - self.taker_fee_bps as f32 * 0.0001)
```
Comment (ckamm) [inline programs/mango-v4/src/state/token_conditional_swap.rs:148]: it'd be more efficient to just store the premium and fees as f32 in the first place - and not as bps

File: programs/mango-v4/src/instructions/token_conditional_swap_create.rs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+use crate::state::*;
+
+#[allow(clippy::too_many_arguments)]
+pub fn token_conditional_swap_create(
+    ctx: Context<TokenConditionalSwapCreate>,
+    token_conditional_swap: TokenConditionalSwap,
+) -> Result<()> {
+    let group = ctx.accounts.group.load()?;
+    require!(
+        group.is_ix_enabled(IxGate::TokenConditionalSwapCreate),
```
Comment (ckamm) [inline programs/mango-v4/src/instructions/token_conditional_swap_create.rs:15]: redundant, already done in accounts struct

File: programs/mango-v4/src/instructions/token_conditional_swap_create.rs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+    let now_ts: u64 = Clock::get()?.unix_timestamp.try_into().unwrap();
+    if token_conditional_swap.is_expired(now_ts) {
+        msg!("Already expired, ignoring");
+        return Ok(());
+    }
+
+    require_gte!(token_conditional_swap.price_premium_bps, 0);
+    require_gte!(token_conditional_swap.maker_fee_bps, 0);
+    require_gte!(token_conditional_swap.taker_fee_bps, 0);
```
Comment (ckamm) [inline programs/mango-v4/src/instructions/token_conditional_swap_create.rs:31]: these checks don't actually help, because the setting is on the group, move later or remove

File: programs/mango-v4/src/state/mango_account.rs [hunk 12]
```diff
@@ -104,7 +105,9 @@ pub struct MangoAccount {
     /// End timestamp of the current expiry interval of the buyback fees amount.
     pub buyback_fees_expiry_timestamp: u64,
 
-    pub reserved: [u8; 208],
+    pub next_conditional_swap_id: u64,
```
Comment (ckamm) [inline programs/mango-v4/src/state/mango_account.rs:108]: call this token_conditional_swap to be consistent

File: programs/mango-v4/src/state/mango_account.rs [hunk 13]
```diff
@@ -629,6 +714,41 @@ impl<
...
+    }
+
+    pub fn token_conditional_swap_by_index(&self, index: usize) -> Result<&TokenConditionalSwap> {
+        require_gt!(self.header().token_conditional_swap_count(), index);
+        Ok(self.token_conditional_swap_by_index_unchecked(index))
+    }
+
+    pub fn token_conditional_swap_by_id(&self, id: u64) -> Result<(usize, &TokenConditionalSwap)> {
+        let index = self
+            .all_token_conditional_swap()
```
Comment (ckamm) [inline programs/mango-v4/src/state/mango_account.rs:730]: do we plural this?

File: programs/mango-v4/src/state/mango_account.rs [hunk 14]
```diff
@@ -1008,6 +1128,23 @@ impl<
...
+        &mut self,
+        index: usize,
+    ) -> Result<&mut TokenConditionalSwap> {
+        let count: usize = self.header().token_conditional_swap_count.into();
+        require_gt!(count, index);
+        let offset = self.header().token_conditional_swap_offset(index);
+        Ok(get_helper_mut(self.dynamic_mut(), offset))
+    }
+
+    pub fn add_token_conditional_swap(&mut self) -> Result<&mut TokenConditionalSwap> {
```
Comment (ckamm) [inline programs/mango-v4/src/state/mango_account.rs:1140]: rename

File: programs/mango-v4/src/state/mango_account.rs [hunk 15]
```diff
@@ -1140,24 +1264,45 @@ impl<
...
             perp_count: new_perp_count,
             perp_oo_count: new_perp_oo_count,
+            token_conditional_swap_count: new_token_conditional_swap_count,
         };
         let old_header = self.header().clone();
         let dynamic = self.dynamic_mut();
 
         // expand dynamic components by first moving existing positions, and then setting new ones to defaults
 
+        // token stop loss
```
Comment (ckamm) [inline programs/mango-v4/src/state/mango_account.rs:1290]: old comment

**Review by ckamm — COMMENTED** [review]
self-reviewed everything but tcs_trigger
